### ISSUE TITLE
`MalformedParametersException` when I add a breakpoint inside a Bean creator method which has parameter

### ISSUE THREAD

**Issue by LiYing2010** [eclipse/openj9#6225]
Java -version output
--------------------
```
openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.3+7)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.14.0, JRE 11 Mac OS X amd64-64-Bit Compressed References 20190418_204 (JIT enabled, AOT enabled)
OpenJ9   - bad1d4d06
OMR      - 4a4278e6
JCL      - 5cc996a803 based on jdk-11.0.3+7)
```

Summary of problem
------------------
I am using `OpenJDK11 with OpenJ9 JVM` on a Mac Book,
when I start my Java App (using Spring Boot), I get a `MalformedParametersException: Invalid parameter name ""` error.

I traced into source code of Spring and JDK itself, looks like Spring try to get the parameter name of method before it invoke it to create a Bean, but native method `java.lang.reflect.Executable#getParameters0` (which rely on JVM) returns Parameters with empty string as parameter name, then `java.lang.reflect.Executable#verifyParameters` failed, and throw this exception.

But the strangest thing is:
this exception only happen if I add a breakpoint inside a Spring Boot Configuration class method which create a Bean, and has parameter for this method.
If I remove breakpoint, everything is fine.
If I add breakpoint inside anywhere else, even inside other Configuration class method, without parameter, everything is fine.

I can't understand why a breakpoint could affect the JVM, maybe somehow class/method/parameter information is messed in DEBUG mode?

My develop environment is:

|    Software or Dependency    | Version |
| ---------- | --- |
| OS |  macOS 10.14.5 |
| JDK     |  openjdk version "11.0.3" <br> openj9-0.14.0 |
| IDE |  InteliJ IDEA ULTIMATE 2019.2 EAP <br> Build #IU-192.5281.24, built on June 19, 2019 <br> *also tried in 2019.1.3 (NOT EAP) edition, but same result |
| Spring Boot | 2.1.0.RELEASE |
| Spring Framework |  5.1.2.RELEASE |

Diagnostic files
----------------
no diagnostic file, but exception stacktrace is:

```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled. 
2019-06-21 11:28:27.828 ERROR [main] o.springframework.boot.SpringApplication:858 : Application run failed 
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'xxxx' defined in file [/xxxx.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'xxxx' defined in file [/xxxx.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'xxxx' defined in class path resource [xxxx.class]: Unexpected exception during bean creation; nested exception is java.lang.reflect.MalformedParametersException: Invalid parameter name ""
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:767)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:218)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1308)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1154)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:538)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:498)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.AbstractBeanFactory$$Lambda$259.00000000A6A48020.getObject(Unknown Source)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:846)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:863)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:546)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248)
	at xxxx.Application.main(Application.java:14)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'xxxx' defined in file [xxxx.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'xxxx' defined in class path resource [xxxx.class]: Unexpected exception during bean creation; nested exception is java.lang.reflect.MalformedParametersException: Invalid parameter name ""
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:767)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:218)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1308)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1154)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:538)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:498)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.AbstractBeanFactory$$Lambda$259.00000000A6A48020.getObject(Unknown Source)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:273)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1239)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1166)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:855)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:758)
	... 20 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'xxxx' defined in class path resource [xxxx.class]: Unexpected exception during bean creation; nested exception is java.lang.reflect.MalformedParametersException: Invalid parameter name ""
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:511)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.AbstractBeanFactory$$Lambda$259.00000000A6A48020.getObject(Unknown Source)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:273)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1239)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1166)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:855)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:758)
	... 35 common frames omitted
Caused by: java.lang.reflect.MalformedParametersException: Invalid parameter name ""
	at java.base/java.lang.reflect.Executable.verifyParameters(Executable.java:382)
	at java.base/java.lang.reflect.Executable.privateGetParameters(Executable.java:412)
	at java.base/java.lang.reflect.Executable.getParameters(Executable.java:353)
	at org.springframework.core.StandardReflectionParameterNameDiscoverer.getParameterNames(StandardReflectionParameterNameDiscoverer.java:39)
	at org.springframework.core.PrioritizedParameterNameDiscoverer.getParameterNames(PrioritizedParameterNameDiscoverer.java:55)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:506)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1288)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1127)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:538)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:498)
	... 45 common frames omitted
Disconnected from the target VM, address: '127.0.0.1:51594', transport: 'socket'
```

**Comment by pshipton** [comment]
@gacholio @andrewcraik

**Comment by gacholio** [comment]
https://github.com/eclipse/openj9/blob/6887baea4c98f2820f42071c29de5f7a7ce31642/runtime/jvmti/jvmtiHelpers.c#L973

This is the code which breakpoints a method - it's not copying the method parameter data (the comment indicates that queries should be using the original ROM method).

**Comment by gacholio** [comment]
The fix is likely to call getOriginalROMMethod in more places in the query code (mthutil.c).

**Comment by pshipton** [comment]
Tentatively assigned to 0.15, just in case we have a fix ready next week that can go in.

**Comment by gacholio** [comment]
I have so far been unable to reproduce this problem with a trivial debug test case.

**Comment by LiYing2010** [comment]
> I have so far been unable to reproduce this problem with a trivial debug test case.

hi @gacholio:

Sorry, I forgot to tell you, this error happens to me, every time, 100% reproducible,
but not happen to my colleague using same version of openj9 JVM.
Maybe there is something related to my hardware or something else.
and also, sometime openj9 JVM crash on my mac, I will send diagnostic files if I can reproduce it.

**Comment by pshipton** [comment]
Since we can't reproduce the problem and don't have any diagnostics it seems unlikely to be resolved in the 0.15 release. Moving it out.

**Comment by gacholio** [comment]
From what I can see, calling `Executeable.getParameters` calls `JVM_GetMethodParameters_Impl` which calls `getMethodParametersAsArray` which always seems to return `NULL` in my testing.

I'm not sure how to get a method parameter attribute into the class file - annotating the method arguments (with `@Deprecated` for simplicity in my test) did not seem to produce the attribute block (I compiled the java code with -g as well to be sure).

For the record, I have not tried reproducing the original problem.

**Comment by pshipton** [comment]
@gacholio Did you set the following on the annotation class?
```
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
```

**Comment by gacholio** [comment]
No - I used the standard `@Deprecated`, which is runtime retention.

**Comment by pshipton** [comment]
I'd try making a new annotation like https://github.com/eclipse/openj9/blob/2f295ea91cdd97ff37eb5d164fc63014ea718bc7/test/functional/Java8andUp/src/org/openj9/test/annotation/MethodAnnotation.java

**Comment by gacholio** [comment]
None of that worked - I'm trying something a bit more hardcore to get a class I can test with.

**Comment by gacholio** [comment]
Aha - compiling the breakpointed method with `-parameters` on the `javac` command line allows me to see the issue.

**Comment by gacholio** [comment]
As suspected, using the original ROM method for the queries fixes the issue. There are likely many more instances of this issue.

**Comment by LiYing2010** [comment]
> Aha - compiling the breakpointed method with `-parameters` on the` javac` command line allows me to see the issue.

Yes, we need this option to compile a class, to make the names of parameters stored into class file bytecode,
then Spring can get the parameter names at runtime, find other beans by these names, and pass these beans as parameters to the bean creation method.
That is where this issue happen.

I am very glad you can reproduce it.

**Comment by pshipton** [comment]
I'll move this back into the 0.15 milestone since there is a PR open, although this is not a promise to include a fix in the 0.15 release.
